examples/autoqasm/5_Pulse_programming_and_dynamical_decoupling.ipynb

```diff
@@ -0,0 +1,263 @@
+   "cell_type": "markdown",
+   "id": "eeee0f33",
+   "metadata": {},
+   "source": [
+    "Below prints out the OpenQASM script of the program. The pulse instructions, delay and capture, are in a `cal` block, which stand for \"calibration\". A `cal` block defines the scope for low level hardware instructions. Today, the hardware instructions are pulse instructions."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "3e76fdca",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "OPENQASM 3.0;\n",
+      "defcalgrammar \"openpulse\";\n",
+      "cal {\n",
+      "    delay[10.0us] q0_rf_frame;\n",
+      "    capture_v0(q0_rf_frame);\n",
+      "}\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(idle(0, 10e-6).to_ir())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2b55c594",
+   "metadata": {},
+   "source": [
+    "## Error supression with dynamical decoupling sequences"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "05d92b91",
+   "metadata": {},
+   "source": [
+    "In the `idle` program above, we intentionally add a delay instruction to the program. In more practical programs, qubit idling is inevitable. It often happens when a gate or pulse instruction applies on multiple qubits, but some of the qubit are still under other instructions. The rest of the qubits are idling and waiting for all qubits to be ready. During this idling time, the qubits are subjected to decoherence, resulting in low program fidelity. Dynamical decoupling is a technique to suppress such errors. It includes alternating pairs of X and Y \"$\\pi$ pulse\", a rotation around the Bloch sphere axis with an angle $\\pi$, applying on the qubits during the idling time. These pulses together are equivalent to an identity gate, but alternating \"$\\pi$ pulses\" make some of the decoherence effectively cancel each other out. To learn more about dynamical decoupling, [this blog post](https://aws.amazon.com/blogs/quantum-computing/suppressing-errors-with-dynamical-decoupling-using-pulse-control-on-amazon-braket/) has a detailed explanations and visualization of dynamical decoupling sequences."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "814647c1",
+   "metadata": {},
+   "source": [
+    "First, we create the X and Y $\\pi$ pulse in terms of the native gates of Rigetti Aspen-M-3 device. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "9130f6b8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pi = np.pi\n",
+    "\n",
+    "def x_pulse(qubit: int):\n",
+    "    # Pi pulse apply on X-axis of Bloch sphere\n",
+    "    qubit = f\"${qubit}\"\n",
+    "    rx(qubit, pi)\n",
+    "\n",
+    "def y_pulse(qubit: int):\n",
+    "    # Pi pulse apply on Y-axis of Bloch sphere\n",
+    "    qubit = f\"${qubit}\"\n",
+    "    rz(qubit, -0.5 * pi)\n",
+    "    rx(qubit, pi)\n",
+    "    rz(qubit, +0.5 * pi)"
```

Comment:
Would it be reasonable to declare these as `@aq.gate`? They contain only unitary instructions. It would reduce the duplication in the generated program. e.g.:
```
@aq.gate
def x_pulse(target: aq.Qubit):
```

Reply:
We could, but that would not run on a current hardware